File: examples/ch14/ts/conceptState.ts
// State — canonical conceptual example. Run: npx tsx conceptState.ts
//
// A Context delegates to its current State; each State implements the operation
// and decides the next state. Behavior changes with state, no conditionals.

export interface State {
  handle(ctx: Context): string;
}

export class Active implements State {
  handle(ctx: Context): string {
    ctx.state = new Done();
    return "active->done";
  }
}

export class Done implements State {
  handle(_ctx: Context): string {
    return "done";
  }
}

export class Context {
  state: State = new Active();
  request(): string {
    return this.state.handle(this);
  }
}

// Demo
const ctx = new Context();
console.log(ctx.request()); // active->done
console.log(ctx.request()); // done

File: examples/ch14/py/concept_state.py
"""State — canonical conceptual example (runnable).

A Context delegates to its current State object; each State implements the
operation and decides the next state. Behavior changes with state, with no
conditionals in the Context.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class State(ABC):
    @abstractmethod
    def handle(self, ctx: Context) -> str: ...


class Active(State):
    def handle(self, ctx: Context) -> str:
        ctx.state = Done()
        return "active->done"


class Done(State):
    def handle(self, ctx: Context) -> str:
        return "done"


class Context:
    def __init__(self) -> None:
        self.state: State = Active()

    def request(self) -> str:
        return self.state.handle(self)


if __name__ == "__main__":
    ctx = Context()
    print(ctx.request())  # active->done
    print(ctx.request())  # done
